Biz/Model/LoginAttempt.pm

```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2023 Bivio Software, Inc.  All Rights Reserved.
+package Bivio::Biz::Model::LoginAttempt;
+use strict;
+use Bivio::Base 'Model.RealmBase';
+
+my($_C) = b_use('IO.Config');
+my($_S) = b_use('Type.LoginAttemptState');
+$_C->register(my $_CFG = {
+    lockout_failure_count => $_C->is_test ? 5 : 100,
+});
+
+sub create {
+    my($self, $values) = @_;
+    return $self->SUPER::create({
+        %$values,
+        login_attempt_state => _state($self, $values),
+        # Utils don't have client_addr set
+        ip_address => $self->req->unsafe_get('client_addr') || '0.0.0.0',
+    });
+}
+
+sub handle_config {
+    my(undef, $cfg) = @_;
+    $_CFG = $cfg;
+    return;
+}
+
+sub internal_initialize {
+    my($self) = @_;
+    return $self->merge_initialize_info($self->SUPER::internal_initialize, {
+        version => 1,
+        table_name => 'login_attempt_t',
+        as_string_fields => [qw(realm_id login_attempt_state ip_address)],
+        columns => {
+            login_attempt_id => ['PrimaryId', 'PRIMARY_KEY'],
+            $self->REALM_ID_FIELD => [$self->REALM_ID_FIELD_TYPE, 'NOT_NULL'],
+            creation_date_time => ['DateTime', 'NOT_NULL'],
+            login_attempt_state => ['LoginAttemptState', 'NOT_NULL'],
+            ip_address => ['IPAddress', 'NOT_NULL'],
+        },
+    });
+}
+
+sub reset_failure_count {
+    my($proto, $req) = @_;
+    $proto->new($req)->create({
+        login_attempt_state => $_S->RESET,
+    });
+    return;
+}
+
+sub unsafe_load_last {
+    my($self) = @_;
+    return $self->unsafe_load_first('creation_date_time DESC, login_attempt_id DESC');
+}
+
+sub update {
+    b_die('login attempt record modification not allowed');
+}
+
+sub _state {
+    my($self, $values) = @_;
+    my($state) = $values->{login_attempt_state};
+    return $state
+        if $state->eq_success || $state->eq_reset;
+    b_die('unexpected login_attempt_state=', $state)
+        unless $state->eq_failure;
+    my($fail_count) = 1;
+    $self->new_other('LoginAttempt')->do_iterate(sub {
+        my($it) = @_;
+        return 0
+            if $it->get('login_attempt_state')->eq_success
+            || $it->get('login_attempt_state')->eq_reset;
+        if (++$fail_count >= $_CFG->{lockout_failure_count}) {
+            $state = $_S->LOCKOUT;
+            return 0;
+        }
+        return 1;
+    }, 'creation_date_time DESC, login_attempt_id DESC');
```

Comment:
You do because there could be multiple attempts with the same creation_date_time.

Reply:
But the order doesn't matter in this case. If there are N attempts within the same second, they all get checked. By not including this, your 

There's a very odd condition that is "safe" to code this way. If there are max_attempts -1 (or 2) failed attempts from the last previous attempt AND there was a failure(s) in that same second as the success. You'll get a miscount, but I think that's ok. There can only be so many attempts per second (from a compute availability perspective) so if we miscount and give a few more attempts, it's fine.

NOTE: you need to add an index just on realm_id.

Reply:
> But the order doesn't matter in this case. If there are N attempts within the same second, they all get checked. By not including this, your

Looks like your sentence got cut off.

> There's a very odd condition that is "safe" to code this way. If there are max_attempts -1 (or 2) failed attempts from the last previous attempt AND there was a failure(s) in that same second as the success. You'll get a miscount, but I think that's ok. There can only be so many attempts per second (from a compute availability perspective) so if we miscount and give a few more attempts, it's fine.

It seems like it would make the code less clear to write to this "very odd condition" when we can just sort by both columns and ensure that the sequence is correct.

It also makes the testing harder because you have to account for the possibility of the count being off.

> NOTE: you need to add an index just on realm_id.

When are we querying for attempts without caring about the order? Or is there a reason that I'm not thinking of to have an index only on realm_id?

Reply:
> "By not including this, your"

Not sure what I was thinking. :)

> code less clear to write to this "very odd condition" 

This case can be ignored. It really only matters when the number of failures is very high (and really just near the max). We might lock somebody out with, say, max-2 attempts, but that's ok. Nobody is going to count this, since we don't publish the number of failures allowed (that would be silly).

> we can just sort by both columns 

This has a cost, and with a large table like this one, a big cost, I believe. The index will get large and less usable over time.

You need a few indices: (login_attempt_id), (realm_id), (realm_id, creation_date_time). If you add this final index (login_attempt_id, realm_id, creation_date_time) for sorting, I'm concerned about Postgres choosing the wrong index.

> ensure that the sequence is correct.

Sequences do not have a guaranteed time order. Testing can be ordered by faking the time.

> It also makes the testing harder because you have to account for the possibility of the count being off.

The testing can use fake times, I believe.

> When are we querying for attempts without caring about the order? Or is there a reason that I'm not thinking of to have an index only on realm_id?

realm_id needs an index, because it's a foreign key. It makes checks for deletes on the foreign table faster (otherwise it has to scan the entire table for constraint violations).

Reply:
And, the larger index (login_attempt_id, realm_id, creation_date_time) is going to be large (possibly as large as the table itself), because none of the values align and PG has to keep ordering.

Reply:
> realm_id needs an index, because it's a foreign key.

Ah, MySQL creates indices for foreign keys automatically.